src/kibali/ProtectedResource.cs

```diff
@@ -263,17 +282,161 @@ public string WriteLeastPrivilegeTable(Dictionary<string, Dictionary<string, Has
+
+                // Filter out least privilege entries from the higher privileged permissions to process
+                higherScopes = higherScopes.Where(p => !invalidHigherPrivilegedEntries.Contains((p, string.Empty)));
+            }
+
+            var higher = StringConstants.PermissionNotAvailable;
+            if (permissionPairs.Any())
+            {
+                ProcessPermissionPairs(permissionPairs, leastPrivilegedPermissionPair, higherScopes, ref higher);
+            }
+            else
+            {
+                higher = higherScopes.Any() ? string.Join(", ", higherScopes) : leastPrivilegedPermissionPair.Any() ? StringConstants.PermissionNotAvailable : StringConstants.PermissionNotSupported;
+            }
 
-            var schemeScopes = methodClaims.TryGetValue(scheme, out List<AcceptableClaim> claims) ? claims.OrderByDescending(c => c.Least).Select(c => c.Permission) : permissionsStub;
-            leastPrivilege.TryGetValue(scheme, out HashSet<string> scopes);
-            (var least, var higher) = ExtractScopes(schemeScopes, scopes);
-            return (least, higher);
+            return higher;
         }
 
-        private void PopulateLeastPrivilege(Dictionary<string, Dictionary<string, HashSet<string>>> leastPrivilege, string method, string scheme, HashSet<string> permissions)
+        private IEnumerable<string> GetRequiredLeastPrivileged(IEnumerable<AcceptableClaim> claims)
+        {
+            var leastPrivilegedClaims = claims?.Where(c => c.Least && c.AlsoRequires.Length > 0);
+            if (leastPrivilegedClaims?.Count() > 2)
+            {
+                throw new Exception($"Too many Least Privilege Entries {string.Join(",", leastPrivilegedClaims.Select(c => c.Permission))}");
+            }
+            if (leastPrivilegedClaims?.Count() == 1)
+            {
+                var claim = leastPrivilegedClaims.First();
+                if (claim.AlsoRequires.Length > 1)
+                {
+                    throw new Exception($"The least privileged scope {claim.Permission} requires more than one other scope. Only one of {string.Join(", ", claim.AlsoRequires)} is allowed.");
+                }
+                return [claim.Permission, claim.AlsoRequires.First()];
+            }
+
+            return leastPrivilegedClaims?.Select(c => c.Permission);
+        }
+
+        private static void ProcessPermissionPairs(HashSet<(string, string)> permissionPairs, IEnumerable<string> leastRequired, IEnumerable<string> higherScopes, ref string higher)
+        {
+            var higherPrivilegedPairs = new List<string>();
+            var found = new HashSet<(string, string)>(new OrderedPairEqualityComparer());
+            foreach (var pair in permissionPairs)
+            {
+                var inverse = (pair.Item2, pair.Item1);
+                if (found.Contains(pair) || found.Contains(inverse))
```

Comment:
```suggestion
                if (found.Contains(pair))
```
I believe the `OrderedPairEqualityComparer` contains the logic that already handles that. `A ^ B` is the same as `B ^ A` since _XOR is Commutative_ for the `GetHashCode` and the `Equals` also does the required comparison.